### PR TITLE
[1.4] Implement NPC Happiness!

### PR DESCRIPTION
**Pull request by kittenchilly**
Pretty self-explanatory. Adds a new hook for ModNPC and GlobalNPC called NPCHappiness that allows you to modify NPC biome and neighbor preferences. Also adds ExampleMod stuff for implementation, localization, etc.

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/Mod.cs [hunk 1]
```diff
@@ -414,7 +414,7 @@ public void AddConfig(string name, ModConfig mc)
 		/// </summary>
 		/// <param name="key">The key for the ModTranslation. The full key will be Mods.ModName.key</param>
```
Comment (Steviegt6) [inline patches/tModLoader/Terraria/ModLoader/Mod.cs:415]: > The full key will be Mods.ModName.key

From the looks of it, this isn't true anymore (and should probably be changed). This is just from a quick glance though, tell me if I'm wrong.

Comment (kittenchilly) [reply patches/tModLoader/Terraria/ModLoader/Mod.cs:415]: Oh oops, guess I missed that.

**Review by Solxanich — CHANGES_REQUESTED** [review]
Regarding the localization change:
I would recommend undoing it as, although straight-forward in what you did, it has potentially far cascading consequences that we won't want to check through all.

Instead, I would recommend changing line ~197 in ShopHelper.cs :
`string str = "TownNPCMood_" + NPCID.... blah`
to be appx:
`string str = "TownNPCMood_" + (isModdedNPC(_currentNPC) ? "Mods." : "") + NPCID.... blah`

Realistically, I don't see why we can't just play with _currentNPCBeingTalkedTo and isMod checks in that method to get the correct translation key string instead of changing mod localization key template.

**Review by Chicken-Bones — APPROVED** [review]
Looks solid to me, just that `TODO: documentation` should probably be replaced with a simple comment saying `make the guide like example person`

File: ExampleMod/Content/NPCs/ExamplePerson.cs [hunk 2]
```diff
@@ -138,6 +139,30 @@ public class ExamplePerson : ModNPC
 			return score >= ((right - left) * (bottom - top)) / 2;
 		}
 
+		//Set Example Person's biome and neighbor preferences with the NPCHappiness hook. You can add happiness text and remarks with localization (See an example in ExampleMod/Localization/en-US.lang).
+		public override void NPCHappiness(int primaryPlayerBiome, ref ShopHelper shopHelperInstance, ref bool[] nearbyNPCsByType) {
+			//TODO: Make it prefer ExampleBiome when that happens. And BiomeID.
```
Comment (Solxanich) [inline ExampleMod/Content/NPCs/ExamplePerson.cs:144]: This TODO should be complete as #1205  has-been merged and includes all the stuff to setup for this.
Should be a simple todo to do

File: patches/tModLoader/Terraria/GameContent/ShopHelper.cs.patch [hunk 3]
```diff
@@ -32,3 +32,13 @@
  		}
  
  		private void AddHappinessReportText(string textKeyInCategory, object substitutes = null) {
++			string str;
++			if (_currentNPCBeingTalkedTo.type <= NPCID.Count)
+-			string str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
++				str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
++			else
++				str = "Mods." + _currentNPCBeingTalkedTo.ModNPC.Mod.Name + "." + _currentNPCBeingTalkedTo.ModNPC.Name + "." + "TownNPCMood";
++
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/GameContent/ShopHelper.cs.patch:41]: Cache search result, use interpolated strings, using a local for modnpc would also be nice.

File: patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs [hunk 4]
```diff
@@ -522,6 +523,16 @@ public virtual void DrawBehind(NPC npc, int index)
...
+		/// <summary>
+		/// Allows you to set an NPC's biome preferences and nearby npc preferences for the NPC happiness system. Recommended to only be used with NPCs that have shops.
+		/// </summary>
+		/// <param name="npc"></param>
+		/// <param name="shopHelperInstance"></param>
+		/// <param name="primaryPlayerBiome"></param>
+		/// <param name="nearbyNPCsByType"></param>
+		public virtual void NPCHappiness(NPC npc, int primaryPlayerBiome, ref ShopHelper shopHelperInstance, ref bool[] nearbyNPCsByType) {
+		}
+
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:535]: Document parameters. What's with the last two parameters being by ref? The last parameter could be a readonlyspan.

**Review by Mirsario — CHANGES_REQUESTED** [review]

File: ExampleMod/Localization/en-US.hjson [hunk 5]
```diff
@@ -62,5 +73,28 @@ Mods: {
...
 		}
+		
+		BiomeName: {
+			ExampleSurfaceBiome: Example Surface
+		}
+		
+		#This is what displays during all of an NPC's happiness levels. Note that the localization key for these should always be "*NPCName*.TownNPCMood.*Mood*" or else it won't display properly.
+		ExamplePerson: {
+			TownNPCMood: {
+				Content: I feel pretty fine right now.
```
[the remark this answers is not in the document]
Comment (Solxanich) [reply ExampleMod/Localization/en-US.hjson]: Resolved

File: patches/tModLoader/Terraria/GameContent/ShopHelper.cs.patch (hunk 3 above)
Comment (Solxanich) [reply patches/tModLoader/Terraria/GameContent/ShopHelper.cs.patch]: Resolved exempting caching search result, not understood what to do with that.
Should be sufficient now

File: patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs (hunk 4 above)
Comment (Solxanich) [reply patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs]: Resolved

**Review by Mirsario — CHANGES_REQUESTED** [review]
A summary of my problems with the current (Vanilla) design for happiness definition:
- The if checks for `nearbyNPCsByType` and `primaryPlayerBiome` are pretty raw boilerplate. Have a look: 
image

- The nature of vanilla's `(Like/Dislike/Love/Hate)(NPC/Biome)` methods results in inability for mods to properly cancel out each other's effects, or do the same thing without conflicts. For example, if 3 mods make the same NPC like another same NPC, then the effect of those calls on `_currentPriceAdjustment` will be tripled, as anyone can see from looking at their code: image

I would heavily prefer if we redo the vanilla system to be more declarative.

**Comment by Mirsario** [comment]
Solved my above concerns with a redesign of the PR. Here's some extra TODOs:

- [ ] ModifyNPCHappiness:
    - [ ] Should be able to directly modify price scaling? Or declare reasons for modifications to it together with happiness remarks?
    - [ ] nearby npcs should be a ReadOnlySpan<bool>.
    - [ ] ShopHelper likely has its remark-raising methods protected and is utterly useless. The hook should be able to modify remarks and add new ones, but I can't picture the best place for that.

**Comment by Solxanich** [comment]
PR Priority : Release. Rationale: minimize bug submissions due to happiness button on town npcs not working since its always present
PR target scope: Allow modders to implement localization text for npc happiness, and basic vanilla relationships.

PR scope change request: Allow custom relationships for happiness outside of vanilla presets of Like/Love/Hate. Priority: either 'Do at some point' or 'Post Release' depending on modder demand. Rationale: may break parts of current implementation, but is a feature addition with insufficient drivers relative to key Release items.
PR scope change: rejected. Rationale: Lower priority scope addition represents scope creep; defer scope to a future PR assuming #2073. Captured deferred scope in #2083 

Mergeability:
Tested: Y
Peer Reviewed: Y          Dev Reviewed: Y
Outstanding Actions: N

**Pull request merged**